conference/src/main/java/com/kyriba/conference/management/service/ScheduleServiceImpl.java

```diff
@@ -0,0 +1,120 @@
+import com.kyriba.conference.management.domain.exception.LinkedEntityNotFound;
+import com.kyriba.conference.management.domain.exception.TimeslotIsBooked;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Optional;
+import java.util.stream.Stream;
+
+
+@Service
+public class ScheduleServiceImpl implements ScheduleService
+{
+  private final PresentationRepository presentationRepository;
+
+  private final HallRepository hallRepository;
+
+
+  @Autowired
+  public ScheduleServiceImpl(PresentationRepository presentationRepository,
+                             HallRepository hallRepository)
+  {
+    this.presentationRepository = presentationRepository;
+    this.hallRepository = hallRepository;
+  }
+
+
+  @Override
+  public Iterable<Presentation> getSchedule()
+  {
+    return presentationRepository.findAll();
+  }
+
+
+  @Transactional
+  @Override
+  public Long addPresentation(PresentationRequest presentationRequest) throws LinkedEntityNotFound
+  {
+    Optional<Hall> hall = hallRepository.findById(presentationRequest.getHall());
+    if (hall.isPresent()) {
+      TopicDto topic = presentationRequest.getTopic();
+      Presentation presentation = Presentation.builder()
+          .hall(hall.get())
+          .topic(new Topic(topic.getTitle(), topic.getAuthor()))
+          .startTime(presentationRequest.getStartTime())
+          .endTime(presentationRequest.getEndTime())
+          .build();
+      presentation.validate();
+      validateTimeIntersection(presentation);
+
+      return presentationRepository.save(presentation).getId();
+    }
+    else {
+      throw new LinkedEntityNotFound("Hall not found");
+    }
+  }
+
+
+  private void validateTimeIntersection(Presentation presentation)
+  {
+    try (Stream<Presentation> intersections = presentationRepository
+        .findByStartTimeBetweenOrEndTimeBetween(presentation.getStartTime(), presentation.getEndTime(),
+            presentation.getStartTime(), presentation.getEndTime())) {
+      if (intersections.anyMatch(pr -> !pr.getId().equals(presentation.getId()))) {
+        throw new TimeslotIsBooked("Timeslot is booked");
+      }
+    }
+  }
+
+
+  @Override
+  public Optional<Presentation> getPresentation(Long id)
+  {
+    return presentationRepository.findById(id);
+  }
+
+
+  @Transactional
+  @Override
+  public void updatePresentation(Long id, PresentationRequest presentationRequest) throws EntityNotFound,
+      LinkedEntityNotFound
+  {
+    Optional<Presentation> presentation = presentationRepository.findById(id);
+    if (presentation.isPresent()) {
+      Presentation updated = presentation.get();
+
+      Hall hall = hallRepository.findById(presentationRequest.getHall())
+          .orElseThrow(() -> new LinkedEntityNotFound("Hall not found"));
+      updated.setHall(hall);
+      updated
```

Comment:
Move business logic to the entity

Reply:
there are call to repository, should I make this calls in the service and just give the result to the entity?